crates/doc/src/heap_de.rs

```diff
@@ -0,0 +1,141 @@
+use super::{BumpStr, BumpVec, HeapField, HeapNode};
+
+use serde::de;
+
+impl<'alloc> HeapNode<'alloc> {
+    /// from_serde deserializes a HeapNode from a serde Deserializer, using a
+    /// provided bump allocator and string de-duplicator.
+    ///
+    /// ```
+    /// use doc::HeapNode;
+    ///
+    /// let alloc = HeapNode::new_allocator();
+    /// let mut deser = serde_json::Deserializer::from_str(r#"{"hello": "world", "one": 2}"#);
+    /// HeapNode::from_serde(&mut deser, &alloc).unwrap();
+    /// ```
+    pub fn from_serde<'de, 'dedup, D>(
+        deser: D,
+        alloc: &'alloc bumpalo::Bump,
+    ) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        deser.deserialize_any(HeapDocVisitor { alloc })
+    }
+}
+
+struct HeapDocVisitor<'alloc> {
+    alloc: &'alloc bumpalo::Bump,
+}
+
+impl<'alloc, 'de> de::Visitor<'de> for HeapDocVisitor<'alloc> {
+    type Value = HeapNode<'alloc>;
+
+    fn expecting(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(fmt, "any value")
+    }
+
+    fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(HeapNode::Bool(v))
+    }
+
+    fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        if v < 0 {
+            Ok(HeapNode::NegInt(v))
+        } else {
+            Ok(HeapNode::PosInt(v as u64))
+        }
+    }
+
+    fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(HeapNode::PosInt(v))
+    }
+
+    fn visit_f64<E>(self, v: f64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(HeapNode::Float(v))
+    }
+
+    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
```

Comment:
I'm wondering whether there might be an easy optimization to avoid copying string data in the case where the JSON deserializer has already allocated a new `String`. At a minimum, I'd expect the deserializer to allocate whenever it needs to parse escape sequences. There's the [`visit_string` function](https://docs.rs/serde/latest/serde/de/trait.Visitor.html#method.visit_string). I don't know the serde-json internals well enough to know that this is worth doing, and I'm also not really sure how much there is to be gained from constructing a `BumpStr` from an owned `String`, so feel free to ignore if that doesn't make sense.

Reply:
> I'd expect the deserializer to allocate whenever it needs to parse escape sequences

It does, but typically in a re-used scratch buffer. `simd-json` will actually mutate the input buffer in-place.

> I'm also not really sure how much there is to be gained from constructing a BumpStr from an owned String
 
You can't, because a `BumpStr` invariant is that its buffer is allocated into the bump allocator chunk. This is what allows it to not implement proper Drop behavior without leaking (because we just through the whole bump allocator and all its chunks out).

We also need to account for that String's allocation in the heap, which would be lost if it were outside of the bump allocator chunks.